Virtual Datasets

frcroth opened the pull request:
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Upload dataset
- Explore remote dataset
- View datasets
- Create, update and save annotations
- Update datasets in WK
- Delete datasets
- Dataset composition
- Zarr streaming
- Zarr streaming via Legacy Routes
- Make changes to JSON and check if they are registered in DB
- Interact with the datasets using the libs/voxelytics/worker

### Implementation notes
DataSource Id is still used internally in the datastore for various caches, also in the binary data handling.
Handling of real disk data sources is still done via orgid and datasetDirectoryName (e.g., uploading, storage size)
Everything else should use dataset ids.

### TODOs:
- [x] WKW datasets can be accessed via the wkDataset routes
- [x] Frontend uses datasetId routes
- [x] Old routes are removed / use the datasetId routes
- [x] Dataset Composition should work
	- [x] Compose frontend
- [x] "Delete dataset on disk" does not necessarily make sense anymore
	- [x] Dataset deletion should set deleted in DB and remove on disk if something is there. 
	- Follow-up #8803 
- [x] Test upload
- [x] Test compose (or rather debug)
- [x] Add a "isVirtual" column to datasets?
- [x] Refactor: Surely we do not need toAbstractLayer and AbstractLayer.from
- [x] Make WKW inherit DataLayerWithMagLocator?
- [x] Frontend schema of datasource properties does not match backend schema (wkwResolutions expected for wkw datasets)
- [x] Test Remote fallback layer
- [x] remove lengthOfUnderlyingCubes

### Issues:
- fixes #8384 

### Followups:
- Readable URIs for zarr routes #8812
- Test remote datasets that are recreated when refreshing schema. #8813
- Protocol Changes #8814
	- Deletion in WK -> WK asks datastore to delete on disk if not virtual
	- Do we need to call register datasets at the datastore? The datastore does not need to be involved in creation of virtual datasets
- Can we make datasetDirectoryName optional?
- make the datastore property optional?
- remove DataSourceRepository entirely, and have the checkInbox function report to wk directly

------
(Please delete unneeded items, merge only when none are left open)
- [x] Added changelog entry (create a `$PR_NUMBER.md` file in `unreleased_changes` or use `./tools/create-changelog-entry.py`)
- [x] Added migration guide entry if applicable (edit the same file as for the changelog)
- [x] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

frcroth commented:
@MichaelBuessemeyer Maybe you could look at the frontend changes? Not sure to assign you as a reviewer here, since I would also be happy if you see something you just change it since I do not know about the frontend anyway

fm3 reviewed (changes requested):
Great stuff, thanks for pushing this through!

I think it is already very close to where we want to go :-) Please have a look at my comments below.

I think one thing that we need to be careful about is the changed format written for WKW datasets (mags instead of wkwResolutions). We need to make sure that if we write datasource-properties.json files like that, it doesn’t break important workflows. Since we don’t keep the `cubeLength`, we cannot simply write out the old format. So this probably means we need to adapt downstream clients.

If I understand correctly, virtual and composed datasets still have a datasetDirectoryName. What is this value then? Is it used for something? Could it become optional and be None for those datasets? (Can be a follow-up, though)

Also, in a follow-up we might want to make the datastore property optional as well, so that any datastore can serve the remote data (e.g. one deployed in the region of the user). Do you think there are important spots still that rely on there being a specific datastore registered for virtual datasets?

Also, I wonder if we could get rid of the stateful DataSourceRepository entirely, and have the checkInbox function report to wk directly). Everything that is available from it should also be readable using the DatasetCache, right?

frcroth commented on app/models/dataset/Dataset.scala:
```diff
@@ -517,6 +519,15 @@ class DatasetDAO @Inject()(sqlClient: SqlClient, datasetLayerDAO: DatasetLayerDA
+                       FROM webknossos.datasets_
+                       WHERE isVirtual
+                       AND $accessQuery""".as[ObjectId])
+    } yield rList.toList
```
yes that makes sense to me

frcroth commented on app/models/dataset/DatasetService.scala:
```diff
@@ -114,7 +139,9 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
       dataSource: InboxDataSource,
-      publication: Option[ObjectId] = None
+      publication: Option[ObjectId] = None,
+      status: Option[String] = None,
```
True, I used this before introducing isVirtual

frcroth commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/dataformats/layers/WKWDataLayers.scala:
```diff
@@ -32,7 +31,7 @@ trait WKWLayer extends DataLayer {
   def wkwResolutions: List[WKWResolution]
 
-  def mags: List[MagLocator] = wkwResolutions.map(_.toMagLocator)
+  def defaultCubeSize = 1024
```
Yes, wkwResolutions I agree, but lengthOfUnderlyingCubes seems to be used (it is set to 32 in VolumeTracingLayer and used for the cube size of data requests in e.g., fallback bucket data

frcroth commented on app/models/dataset/explore/WKExploreRemoteLayerService.scala:
```diff
@@ -105,17 +105,21 @@ class WKExploreRemoteLayerService @Inject()(credentialService: CredentialService
-      userToken <- bearerTokenService.createAndInitDataStoreTokenForUser(user)
-      _ <- client.addDataSource(organization._id, datasetName, dataSource, folderId, userToken)
+      datasetId <- datasetService.createVirtualDataset(
+        dataSource.id.directoryName,
```
If everything was virtual, we would not need it. But for virtual datasets, we could use it as a URL safe name (for notion style routes).

frcroth commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala:
```diff
@@ -782,11 +682,31 @@ class DataSourceController @Inject()(
     }
   }
 
+  private def refreshDataSource(datasetId: ObjectId)(implicit tc: TokenContext): Fox[DataSource] =
+    for {
+      dataSourceInDB <- datasetCache.getById(datasetId) ~> NOT_FOUND
+      dataSourceId = dataSourceInDB.id
+      dataSourceFromDir <- Fox.runIf(
+        dataSourceService.existsOnDisk(dataSourceId.organizationId, dataSourceId.directoryName)) {
+        dataSourceService
+          .dataSourceFromDir(
+            dataSourceService.dataBaseDir.resolve(dataSourceId.organizationId).resolve(dataSourceId.directoryName),
+            dataSourceId.organizationId)
+          .toUsable
+          .toFox
+      }
+      _ <- dataSourceFromDir match {
+        case Some(ds) => dsRemoteWebknossosClient.updateDataSource(ds, datasetId)
+        case _        => Fox.successful(())
+      }
+      dataSource <- datasetCache.getById(datasetId) ~> NOT_FOUND
+    } yield dataSource
+
```
Right, I forgot the invalidate call

fm3 commented on app/controllers/WKRemoteDataStoreController.scala:
```diff
@@ -270,6 +295,53 @@ class WKRemoteDataStoreController @Inject()(
+          oldDataSource <- datasetService.fullDataSourceFor(dataset)
+          oldPaths = oldDataSource.toUsable.map(_.allExplicitPaths).getOrElse(List.empty)
+          newPaths = request.body.allExplicitPaths
+          _ <- Fox.fromBool(newPaths.forall(oldPaths.contains))
```
should probably get an error message via `?~>`

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DSRemoteWebknossosClient.scala:
```diff
@@ -197,4 +227,11 @@ class DSRemoteWebknossosClient @Inject()(
       .getWithJsonResponse[GenericDataSource[DataLayer]] ?~> "Failed to get data source from remote webknossos"
+
+  def getDatasetId(organizationId: String, datasetDirectoryName: String): Fox[ObjectId] =
+    rpc(s"$webknossosUri/api/datastores/$dataStoreName/findDatasetId")
+      .addQueryString("key" -> dataStoreKey)
+      .addQueryString("organizationId" -> organizationId)
+      .addQueryString("datasetDirectoryName" -> datasetDirectoryName)
+      .getWithJsonResponse[ObjectId] ?~> "Failed to get dataset id from remote webknossos"
```
This should be cached on the datastore side to save us some requests.

fm3 reviewed (commented):
Backend code looking good, thanks for the latest changes! I’ll do some testing tomorrow.

Could you write follow-up issues for the discussed ideas? Mostly to switch the protocol to frontend<->wk instead of frontend<->datastore where applicable

fm3 reviewed (changes requested):
Latest commit on caching looks good :)

I did some testing and dataset reading and annotating works well :) However, with more specific use cases I encountered a few errors as well. Please have a look at these, feel free to ask me if there is something you can’t reproduce or if my description is too vague.

- [ ] After downloading an annotation and trying to reupload it I get an error `Failed RPC 5: POST http://localhost:9000/tracings/volume/a5a485fd-dad4-46a8-bdb4-2f90868eb7c1/initialData. Response: 400 '{"messages":[{"error":"Failed RPC 8: GET http://localhost:9000/api/tracingstores/localhost/datasetId. Response: 400 '{\"messages\":[{\"error\":\"Annotation could not be found\"}]}'"}]}'`

- [ ] When trying to self-stream an annotation (added its zarr link `http://localhost:9000/data/annotations/zarr/Q6Gp91pYHGdj6jcZ` as remote dataset, note that it’s not layer-specific), I can explore it but I don’t see any data in the result. Not sure if this problem was introduced in this PR. Logging says `Could not read header at .zarray`

- [ ] When opening a dataset settings view `http://localhost:9000/datasets/68888251d001008f02a0d46c/edit/data` I immediately get `Failed RPC 64: PUT http://localhost:9000/api/datastores/localhost/datasources/68888251d001008f02a0d46c. Response: 400 '{"messages":[{"error":"New mag paths must be a subset of old mag paths"}]}'`. The zarr3 dataset was scanned from disk and has a datasource-properties.json with paths in the style of `"path": "./color/1",`.

- [ ] Opening the settings view for a virtual dataset works. When I want to change its rotation matrix and hit save, there is no error, but on the next reload, it’s zero again.

- [ ] In the worker (I tried `infer_neurons`) the publish task fails. I’m getting `/data/v9/triggers/reload/sample_organization/myDataset' [Cannot parse parameter datasetId as ObjectId: myDataset]` . I think we need to add this route in the legacy controller for backwards compatibility. After this request it also tries to `GET http://localhost:9000/data/v9/zarr/sample_organization/myDataset/datasource-properties.json`, which also fails with the same error. I’m not sure why because this route does appear in the legacy controller. Maybe you could have a look here too.

frcroth commented:
>     * [ ]  After downloading an annotation and trying to reupload it I get an error `Failed RPC 5: POST http://localhost:9000/tracings/volume/a5a485fd-dad4-46a8-bdb4-2f90868eb7c1/initialData. Response: 400 '{"messages":[{"error":"Failed RPC 8: GET http://localhost:9000/api/tracingstores/localhost/datasetId. Response: 400 '{\"messages\":[{\"error\":\"Annotation could not be found\"}]}'"}]}'`

Fixed that by also including the datasetId in the temporary store for data sources.

>     * [ ]  When trying to self-stream an annotation (added its zarr link `http://localhost:9000/data/annotations/zarr/Q6Gp91pYHGdj6jcZ` as remote dataset, note that it’s not layer-specific), I can explore it but I don’t see any data in the result. Not sure if this problem was introduced in this PR. Logging says `Could not read header at .zarray`

I can see data, though it is always in the lower left corner.
 
>     * [ ]  When opening a dataset settings view `http://localhost:9000/datasets/68888251d001008f02a0d46c/edit/data` I immediately get `Failed RPC 64: PUT http://localhost:9000/api/datastores/localhost/datasources/68888251d001008f02a0d46c. Response: 400 '{"messages":[{"error":"New mag paths must be a subset of old mag paths"}]}'`. The zarr3 dataset was scanned from disk and has a datasource-properties.json with paths in the style of `"path": "./color/1",`.

I changed one thing that may have fixed it, at least I can not reproduce it right now.
 
>     * [ ]  Opening the settings view for a virtual dataset works. When I want to change its rotation matrix and hit save, there is no error, but on the next reload, it’s zero again.

I have reproduced this, though it seems to be an issue with the frontend. After saving it and reloading, the transformations are reported to the frontend, but not shown in the UI.

>     * [ ]  In the worker (I tried `infer_neurons`) the publish task fails. I’m getting `/data/v9/triggers/reload/sample_organization/myDataset' [Cannot parse parameter datasetId as ObjectId: myDataset]` . I think we need to add this route in the legacy controller for backwards compatibility. After this request it also tries to `GET http://localhost:9000/data/v9/zarr/sample_organization/myDataset/datasource-properties.json`, which also fails with the same error. I’m not sure why because this route does appear in the legacy controller. Maybe you could have a look here too.

Added the route. The legacy routes weren't actually being used because the generic /v9/ matching was done before any of the specific routes.

fm3 reviewed (commented):
Sweet, most of the problems are gone with the newest commits :)

One last thing: the triggers/reload route is currently used by the voxelytics publish task to signal that voxelytics has now placed the dataset on disk and webknossos should scan it from there.

We plan on changing the protocol for the publish task, introducing a new finishUpload call instead. However, for the moment it is necessary that the v9 triggers/reload call scans the directory at the given directoryName and makes that dataset usable.

Let me know if I can support you further on this

fm3 reviewed (approved):
Let’s go!
